lib/web_ui/lib/src/engine/semantics/text_field.dart

```diff
@@ -341,75 +375,103 @@ class TextField extends RoleManager {
-      editingState.applyTextToDomElement(editableElement);
+      // Focus on semantics element before removing the editable element, so that
+      // the user can continue navigating the page with the assistive technology.
+      semanticsObject.element.focus();
+      editableElement = null;
+    }));
+  }
 
-      if (_hasFocused && domDocument.activeElement == editableElement) {
-        // Unlike `editableElement.focus()` we don't need to schedule `blur`
-        // post-update because `document.activeElement` implies that the
-        // element is already attached to the DOM. If it's not, it can't
-        // possibly be focused and therefore there's no need to blur.
-        editableElement.blur();
-      }
-      _hasFocused = false;
+  @override
+  void update() {
+    if (editableElement != null) {
+      activeEditableElement.style
+        ..width = '${semanticsObject.rect!.width}px'
+        ..height = '${semanticsObject.rect!.height}px';
     }
 
-    if (needsDomFocusRequest) {
-      // Schedule focus post-update to make sure the element is attached to
-      // the document. Otherwise focus() has no effect.
-      semanticsObject.owner.addOneTimePostUpdateCallback(() {
-        if (domDocument.activeElement != editableElement) {
-          editableElement.focus();
+    if (semanticsObject.hasFocus) {
+      // Ignore the update if editableElement has not been created yet.
+      // On iOS Safari, when the user dismisses the keyboard using the 'done' button,
+      // we recieve a `blur` event from the browswer and a semantic update with
+      // [hasFocus] set to true from the framework. In this case, we ignore the update
+      // and wait for a tap event before invoking the iOS workaround and creating
+      // the editable element.
+      if (editableElement != null) {
+        if (flutterViewEmbedder.glassPaneShadow!.activeElement != editableElement) {
+          semanticsObject.owner.addOneTimePostUpdateCallback(() {
+            activeEditableElement.focus();
+          });
         }
-      });
+        SemanticsTextEditingStrategy.instance.activate(this);
+        _setEditingState();
+      }
+    } else if (editableElement != null &&
+        flutterViewEmbedder.glassPaneShadow!.activeElement == editableElement) {
+      if (!isIosSafari) {
+        SemanticsTextEditingStrategy.instance.deactivate(this);
+        // Only apply text, because this node is not focused.
+        _applyTextFromEditingState();
+      }
+      activeEditableElement.blur();
+    }
+
+    final DomElement element = editableElement ?? semanticsObject.element;
+    if (semanticsObject.hasLabel) {
+      element.setAttribute(
+        'aria-label',
+        semanticsObject.label!,
+      );
+    } else {
+      element.removeAttribute('aria-label');
     }
   }
 
   @override
   void dispose() {
-    editableElement.remove();
+    _positionInputElementTimer?.cancel();
+    _positionInputElementTimer = null;
+    activeEditableElement.remove();
```

Comment:
Are we guaranteed to have an `activeEditableElement`? If not, this can throw. I can imagine a text field to never be focused and disposed. Also, we might want to have a test for disposing both focused and unfocused text fields.

Reply:
Good point! Done!